core/src/cache.rs

```diff
@@ -1037,36 +1037,40 @@ impl CacheHub {
         Ok(())
     }
 
-    /// Applies a custom transform to an input and its imports, leaving them in the same state as
-    /// before. Requires that the input has been parsed. In order for the transform to apply to
-    /// imports, they need to have been resolved.
+    /// Applies a custom transform to an input and its imports. Requires that source has been typechecked.
+    ///
+    /// If multiple invocations of `custom_transform` are needed, you must supply `transform_id` with
+    /// with a number higher than that of all previous invocations.
     pub fn custom_transform<E>(
         &mut self,
         file_id: FileId,
+        transform_id: usize,
         f: &mut impl FnMut(&mut CacheHub, RichTerm) -> Result<RichTerm, E>,
     ) -> Result<(), CacheError<E>> {
         match self.terms.entry_state(file_id) {
-            Some(state) if state >= EntryState::Parsed => {
-                if state < EntryState::Transforming {
+            Some(state) if state <= EntryState::CustomTransformed{ transform_id: transform_id } => {
```

Comment:
I think this needs to go back to the older version: we need to error if we haven't reached `EntryState::Parsed` yet.

Reply:
`EntryState::Parsed` is the first state so I thought thats always true - is it possible to have an entry in the terms cache that isnt parsed?

Reply:
Oh, indeed, sorry. I wonder why we have so many other comparisons to `EntryState::Parsed` then. Maybe there used to be other states?

Reply:
That would be my guess! It works regardless as I imagine unparsed meant there was no entry in the term cache, and `_ => Err(CacheError::NotParsed),` would match a `None`.